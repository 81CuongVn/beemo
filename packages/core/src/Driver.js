/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { Plugin } from 'boost';
import merge from 'lodash/merge';
import Options, { array, bool, number, object, string, union } from 'optimal';

import type { EventListener } from 'boost';
import typeof Yargs from 'yargs';
import type { Execution } from './types';

type DriverOptions = {
  args: string[],
  dependencies: string[],
  env: { [key: string]: string },
};

type Metadata = {
  bin: string,
  configName: string,
  configOption: string,
  dependencies: string[],
  description: string,
  helpOption: string,
  title: string,
  useConfigOption: boolean,
};

export default class Driver extends Plugin<DriverOptions> {
  metadata: Metadata = {
    bin: '',
    configName: '',
    configOption: '',
    dependencies: [],
    description: '',
    helpOption: '',
    title: '',
    useConfigOption: false,
  };

  constructor(options?: Object = {}) {
    super(options);

    this.options = new Options(this.options, {
      args: array(string()),
      dependencies: array(string()),
      env: object(union([
        bool(),
        number(),
        string(),
      ])),
    }, {
      name: this.constructor.name,
    });
  }

  /**
   * Setup additional command options.
   */
  bootstrapCommand(command: Yargs) {}

  /**
   * Format the configuration file before it's written.
   */
  formatFile(data: Object): string {
    return JSON.stringify(data, null, 2);
  }

  /**
   * Handle command failures according to this driver.
   */
  handleFailure(error: Execution) {
    const { stderr, stdout } = error;
    const out = (stderr || stdout).trim();

    // Integration debugging
    // this.tool.logError(JSON.stringify(error));

    if (out) {
      this.tool.logError(out);
    }
  }

  /**
   * Handle successful commands according to this driver.
   */
  handleSuccess(response: Execution) {
    const out = response.stdout.trim();

    // Integration debugging
    // this.tool.log(JSON.stringify(response));

    if (out) {
      this.tool.log(out);
    }
  }

  /**
   * Merge multiple configuration objects.
   */
  mergeConfig(prev: Object, next: Object): Object {
    return merge(prev, next);
  }

  /**
   * Easily bind event listeners.
   */
  on(eventName: string, listener: EventListener): this {
    this.tool.on(eventName, listener);

    return this;
  }

  /**
   * Set metadata about the binary/executable in which this driver wraps.
   */
  setMetadata(metadata: Object): this {
    this.metadata = new Options(metadata, {
      bin: string().match(/^[-a-z0-9]+$/),
      configName: string().required(),
      configOption: string('--config'),
      dependencies: array(string()),
      description: string().empty(),
      helpOption: string('--help'),
      title: string(),
      useConfigOption: bool(),
    }, {
      name: this.constructor.name,
    });

    return this;
  }
}
