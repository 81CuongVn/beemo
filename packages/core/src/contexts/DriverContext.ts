/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import Context from './Context';
import Driver from '../Driver';
import { Arguments, Argv } from '../types';

export default class DriverContext extends Context {
  configPaths: string[] = [];

  driverName: string = '';

  drivers: Set<Driver<any>> = new Set();

  parallelArgv: Argv[] = [];

  primaryDriver: Driver<any>;

  workspaceRoot: string = '';

  workspaces: string[] = [];

  constructor(args: Arguments, driver: Driver<any>, parallelArgv: Argv[] = []) {
    super(args);

    this.driverName = driver.name;
    this.parallelArgv = parallelArgv;
    this.primaryDriver = driver;

    // Make the context available in the current driver
    driver.context = this;
  }

  /**
   * Add a driver as a dependency.
   */
  addDriverDependency(driver: Driver<any>): this {
    if (driver instanceof Driver) {
      this.drivers.add(driver);
    } else {
      throw new TypeError('Invalid driver. Must be an instance of `Driver`.');
    }

    return this;
  }

  /**
   * Add a parallel command with additional argv.
   */
  addParallelCommand(argv: Argv): this {
    this.parallelArgv.push(argv);

    return this;
  }

  /**
   * Find a configuration path by file name.
   */
  findConfigByName(fileName: string): string | undefined {
    return this.configPaths.find(path => path.endsWith(fileName));
  }
}
