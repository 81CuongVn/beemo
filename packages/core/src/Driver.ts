/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Plugin, EventListener } from 'boost';
import merge from 'lodash/merge';
import optimal, { array, bool, number, object, shape, string, union, Blueprint } from 'optimal';
import DriverContext from './contexts/DriverContext';
import {
  Argv,
  BeemoConfig,
  DriverCommandOptions,
  DriverOptions,
  DriverMetadata,
  Execution,
} from './types';

export const STRATEGY_REFERENCE = 'reference';
export const STRATEGY_COPY = 'copy';

export default class Driver<T> extends Plugin<DriverOptions> {
  command: DriverCommandOptions = {};

  // @ts-ignore Set after instantiation
  config: T;

  // @ts-ignore Set after instantiation
  context: DriverContext;

  // @ts-ignore Set after instantiation
  metadata: DriverMetadata;

  constructor(options: Partial<DriverOptions> = {}) {
    super(options);

    this.options = optimal(
      options,
      {
        args: array(string()),
        dependencies: array(string()),
        env: object(union([bool(), number(), string()])),
      },
      {
        name: this.constructor.name,
      },
    );
  }

  /**
   * Format the configuration file before it's written.
   */
  formatConfig(data: T): string {
    const content = JSON.stringify(data, null, 2);

    if (this.metadata.configName.endsWith('.js')) {
      return `module.exports = ${content};`;
    }

    return content;
  }

  /**
   * Return a list of user defined arguments.
   */
  getArgs(): Argv {
    return [...this.options.args];
  }

  /**
   * Return a list of dependent drivers.
   */
  getDependencies(): string[] {
    return [
      // Always required; configured by the driver
      ...this.metadata.dependencies,
      // Custom; configured by the consumer
      ...this.options.dependencies,
    ];
  }

  /**
   * Return a list of supported CLI options.
   */
  getSupportedOptions(): string[] {
    return [];
  }

  /**
   * Handle command failures according to this driver.
   */
  handleFailure(error: Execution) {
    const { stderr, stdout } = error;
    const out = (stderr || stdout).trim();

    // Integration debugging
    // this.tool.logError(JSON.stringify(error));

    if (out) {
      this.tool.logError(out);
    }
  }

  /**
   * Handle successful commands according to this driver.
   */
  handleSuccess(response: Execution) {
    const out = response.stdout.trim();

    // Integration debugging
    // this.tool.log(JSON.stringify(response));

    if (out) {
      this.tool.log(out);
    }
  }

  /**
   * Merge multiple configuration objects.
   */
  mergeConfig(prev: T, next: T): T {
    return merge(prev, next);
  }

  /**
   * Easily register events in the tool.
   */
  on(eventName: string, listener: EventListener): this {
    this.tool.on(eventName, listener);

    return this;
  }

  /**
   * Setup additional command options.
   */
  setCommandOptions(options: DriverCommandOptions): this {
    const blueprint: Blueprint = {};

    Object.keys(options).forEach(key => {
      blueprint[key] = shape({
        alias: union([string(), array(string())], ''),
        description: string().required(),
      });
    });

    this.command = optimal(options, blueprint, {
      name: this.constructor.name,
    });

    return this;
  }

  /**
   * Set metadata about the binary/executable in which this driver wraps.
   */
  setMetadata(metadata: Partial<DriverMetadata>): this {
    this.metadata = optimal(
      metadata,
      {
        bin: string()
          .match(/^[-a-z0-9]+$/)
          .required(),
        configName: string().required(),
        configOption: string('--config').empty(),
        dependencies: array(string()),
        description: string().empty(),
        filterOptions: bool(false),
        helpOption: string('--help'),
        title: string().required(),
        useConfigOption: bool(),
        workspaceStrategy: string(STRATEGY_REFERENCE).oneOf([STRATEGY_REFERENCE, STRATEGY_COPY]),
      },
      {
        name: this.constructor.name,
      },
    );

    return this;
  }
}
