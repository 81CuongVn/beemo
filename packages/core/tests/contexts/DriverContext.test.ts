import DriverContext from '../../src/contexts/DriverContext';
import Driver from '../../src/Driver';
import { MOCK_DRIVER_ARGS } from '../../../../tests/helpers';

describe('Context', () => {
  let context: DriverContext;

  beforeEach(() => {
    context = new DriverContext(MOCK_DRIVER_ARGS, new Driver());
  });

  describe('constructor()', () => {
    it('sets args', () => {
      context = new DriverContext({ ...MOCK_DRIVER_ARGS, workspaces: '*' }, new Driver());

      expect(context.args).toEqual({ ...MOCK_DRIVER_ARGS, workspaces: '*' });
    });

    it('sets driver', () => {
      const driver = new Driver();
      driver.name = 'bar';

      context = new DriverContext(MOCK_DRIVER_ARGS, driver);

      expect(context.primaryDriver).toBe(driver);
      expect(context.driverName).toBe('bar');
    });

    it('adds to driver list', () => {
      const driver = new Driver();
      driver.name = 'bar';

      context = new DriverContext(MOCK_DRIVER_ARGS, driver);

      expect(Array.from(context.drivers)).toEqual([driver]);
    });
  });

  describe('addDriverDependency()', () => {
    it('adds a driver', () => {
      expect(Array.from(context.drivers)).toEqual([context.primaryDriver]);

      const driver = new Driver();

      context.addDriverDependency(driver);

      expect(Array.from(context.drivers)).toEqual([context.primaryDriver, driver]);
    });

    it('errors when not a driver', () => {
      expect(() => {
        // @ts-ignore
        context.addDriverDependency(true);
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('addParallelCommand()', () => {
    it('adds a new command argvs', () => {
      expect(context.parallelArgv).toEqual([]);

      context.addParallelCommand(['--foo', 'bar']);
      context.addParallelCommand(['--baz=123']);

      expect(context.parallelArgv).toEqual([['--foo', 'bar'], ['--baz=123']]);
    });
  });
});
