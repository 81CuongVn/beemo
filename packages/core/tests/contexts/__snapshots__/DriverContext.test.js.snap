// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context addDriverDependency() errors when not a driver 1`] = `"Invalid driver. Must be an instance of \`Driver\`."`;
